### PR TITLE
BAH-4651 | Add. new API for Encounter Match & Resume Decision

### PR DESCRIPTION
**Pull request by sahupadma**
JIRA - https://bahmni.atlassian.net/browse/BAH-4651

**Description**:
- Adds `POST /rest/v1/bahmnicore/bahmniencounter/match-decision` endpoint that returns whether the current session matches an existing open encounter.
- New `BahmniEncounterMatchDecisionService` reuses `EncounterSessionMatcher` for the happy path and adds a diagnostic pass to return actionable no-match reasons (`session_expired`, `provider_mismatch`, `location_mismatch`, `no_active_encounter`, `no_active_visit`).
- New DTOs: `EncounterMatchRequest` and `EncounterMatchResponse` (with nested `Ref` and `MatchDetails`) in `bahmni-emr-api`.

### CODE REVIEW

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchResponse.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+    public String getErrorMessage() {
+        return errorMessage;
+    }
+
+    public void setErrorMessage(String errorMessage) {
+        this.errorMessage = errorMessage;
+    }
+
+    // --- Static nested classes ---
```
Comment (ravinderkabli) [inline bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchResponse.java]: Remove the unncessary comments

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchResponse.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+    public static EncounterMatchResponse error(String errorCode, String errorMessage) {
+        EncounterMatchResponse response = new EncounterMatchResponse();
+        response.status = STATUS_ERROR;
+        response.errorCode = errorCode;
+        response.errorMessage = errorMessage;
+        return response;
+    }
+
+    // --- Getters and setters ---
```
Comment (ravinderkabli) [inline bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchResponse.java]: Remove comment

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+            if (requestedVisitLocation != null && !requestedVisitLocation.equals(encounterVisitLocation)) {
+                logger.debug("Location mismatch for encounter: {}", candidate.getUuid());
+                return EncounterMatchResponse.noMatch(
+                        EncounterMatchResponse.REASON_LOCATION_MISMATCH,
+                        "Encounter exists but is in a different location. A new encounter will be created.");
+            }
+        }
+
+        logger.debug("All diagnostic checks passed. Treating as match.");
+        return buildMatchFoundResponse(candidate);
```
Comment (ravinderkabli) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: blocking: `diagnoseNoMatch` fallback returns `match_found` contradicting the authoritative matcher.

When `EncounterSessionMatcher.findEncounter()` returns `null`, the diagnostic pass runs. If all diagnostic checks pass (session not expired, provider matches, location matches), this line returns `match_found` for an encounter the matcher explicitly excluded. The diagnostic query also passes `null` for encounter type (fetching all types), while the matcher filters by encounter type — so the diagnostic can find encounters of a different type.

The diagnostic should only explain *why* there was no match, not override the matcher's decision. Consider returning `noMatch` with a reason like `"no_active_encounter"` or `"encounter_type_mismatch"` here instead.

File: bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java [hunk 4]
```diff
@@ -46,12 +50,37 @@ public BahmniEncounterController() {
...
+
+    @RequestMapping(method = RequestMethod.POST, value = "/match-decision")
+    @ResponseBody
+    public Map<String, Object> matchDecision(@RequestBody EncounterMatchRequest request) {
+        EncounterMatchResponse response = encounterMatchDecisionService.decideMatch(request);
+        return removeNulls(response);
+    }
+
+    private Map<String, Object> removeNulls(EncounterMatchResponse response) {
+        Map<String, Object> result = new HashMap<>();
```
Comment (ravinderkabli) [inline bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java]: suggestion: Consider using `@JsonInclude(JsonInclude.Include.NON_NULL)` on `EncounterMatchResponse` instead of this manual null-field stripping. 

Refer to org.bahmni.module.referencedata.labconcepts.contract.Resource.java, it is being used there.

This approach: (a) breaks type-safety (callers get an untyped map), (b) must be manually kept in sync when fields are added to `EncounterMatchResponse`, and (c) differs from every other endpoint in this controller which returns typed DTOs. The endpoint could return `EncounterMatchResponse` directly with the Jackson annotation handling null exclusion at the serialization layer.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    private Date getSearchStartDate(Date endDate) {
+        int sessionDuration = getSessionDurationMinutes();
+        Date startDate = DateUtils.addMinutes(endDate, -sessionDuration);
+        if (!DateUtils.isSameDay(startDate, endDate)) {
+            return DateUtils.truncate(endDate, Calendar.DATE);
+        }
+        return startDate;
+    }
+
+    private int getSessionDurationMinutes() {
```
Comment (ravinderkabli) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java:312]: suggestion: This duplicates `EncounterSessionMatcher.getSessionDuration()` — same global property, same default, same fallback pattern. This is the third copy in the codebase (also in `OrderSaveCommandImpl`). Also note a subtle behavioral divergence: this version silently catches `NumberFormatException` while the original lets it propagate.

Consider making `EncounterSessionMatcher.getSessionDuration()` public or package-accessible so this service can call it directly.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+            Map<String, Object> context = new HashMap<>();
+            context.put(EncounterSessionMatcher.PATIENT_PROGAM_UUID, request.getPatientProgramUuid());
+            params.setContext(context);
+        }
+
+        Encounter matchedEncounter;
+        try {
+            matchedEncounter = encounterSessionMatcher.findEncounter(visit, params);
+        } catch (RuntimeException e) {
+            if (e.getMessage() != null && e.getMessage().contains("More than one encounter matches")) {
```
Comment (ravinderkabli) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: suggestion: This detection relies on `e.getMessage().contains("More than one encounter matches")` — a string-coupling dependency on an internal error message in `EncounterSessionMatcher` (line 114). If that message is ever changed, this check silently stops working and the exception becomes an unhandled 500 instead of the structured error response.

Consider introducing a dedicated exception type (e.g., `MultipleEncountersMatchException`) in `EncounterSessionMatcher` and catching that specific type here.

File: bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java [hunk 7]
```diff
@@ -37,7 +40,8 @@ public class BahmniEncounterController extends BaseRestController {
     private EncounterTransactionMapper encounterTransactionMapper;
     private BahmniEncounterTransactionService bahmniEncounterTransactionService;
     private BahmniEncounterTransactionMapper bahmniEncounterTransactionMapper;
-    private static Logger logger = LogManager.getLogger(BahmniEncounterController.class);
+    private EncounterMatchDecisionService encounterMatchDecisionService;
+    private static Logger logger = LoggerFactory.getLogger(BahmniEncounterController.class);
```
Comment (ravinderkabli) [inline bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java]: suggestion: All other controllers in this package (`AdminImportController`, `BahmniDrugOrderController`, `BahmniEncounterModifierController`, etc.) use `org.apache.logging.log4j` (log4j2), not slf4j. This introduces a third logging API into the same package.

Consider using `org.apache.logging.log4j.LogManager` and `org.apache.logging.log4j.Logger` for consistency with the codebase.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/EncounterMatchDecisionService.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package org.bahmni.module.bahmnicore.service;
+
+import org.openmrs.module.bahmniemrapi.encountertransaction.contract.EncounterMatchRequest;
+import org.openmrs.module.bahmniemrapi.encountertransaction.contract.EncounterMatchResponse;
+
+public interface EncounterMatchDecisionService {
```
Comment (ravinderkabli) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/EncounterMatchDecisionService.java]: nit: All other domain service interfaces in this package use the `Bahmni` prefix (`BahmniDrugOrderService`, `BahmniDiagnosisService`, `BahmniObsService`, etc.). Consider renaming to `BahmniEncounterMatchDecisionService` for consistency.

**Review by ravinderkabli — COMMENTED** [review]
## Review for BAH-4651

Good progress on the encounter match decision API. All 8 AC scenarios are implemented and tested. Found a few things to address before merge.

Found 1 blocking issue, 4 suggestions, and 1 nit.

**Acceptance Criteria**: 8/8 implemented (note: API contract uses `providerUuid` (singular) vs JIRA spec `providerUuids` (plural) — needs clarification with the team)

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchResponse.java (hunk 1 above)
Comment (sahupadma) [reply bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchResponse.java]: removed

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchResponse.java (hunk 2 above)
Comment (sahupadma) [reply bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchResponse.java]: removed

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java (hunk 3 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: changed to suggested logic

File: bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java (hunk 4 above)
Comment (sahupadma) [reply bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java]: @JsonInclude(JsonInclude.Include.NON_NULL) is not working as openmrs has its own logic which restricts this annotation so going ahead with this approach only

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java (hunk 5 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java:312]: done changed EncounterSessionMatcher.getSessionDuration() public

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java (hunk 6 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: added new exception

File: bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java (hunk 7 above)
Comment (sahupadma) [reply bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java]: done

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/EncounterMatchDecisionService.java (hunk 8 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/EncounterMatchDecisionService.java]: done

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Collection;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+@Service
```
Comment (sumaztwcode) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: blocking: `@Service` bypasses OpenMRS serviceInterceptors.

All other production services in `bahmnicore-api` are registered via XML `TransactionProxyFactoryBean` in `moduleApplicationContext.xml` with `preInterceptors` pointing to `serviceInterceptors` — the OpenMRS auth/authorization and logging chain. Using plain `@Service` bypasses that chain entirely.

Register this service in `moduleApplicationContext.xml` following the same pattern used by `bahmniDiagnosisService`, `bahmniEncounterTransactionService`, etc.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+        this.locationService = locationService;
+        this.providerService = providerService;
+        this.encounterService = encounterService;
+        this.encounterSessionMatcher = encounterSessionMatcher;
+        this.bahmniVisitLocationService = bahmniVisitLocationService;
+        this.administrationService = administrationService;
+    }
+
+    @Override
+    public EncounterMatchResponse decideMatch(EncounterMatchRequest request) {
```
Comment (sumaztwcode) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java:65]: blocking: Missing `@Transactional(readOnly = true)` — risk of `LazyInitializationException`.

`decideMatch()` makes multiple OpenMRS service calls without a transaction boundary. Each call opens/closes its own Hibernate session, so accessing lazy collections like `encounter.getEncounterProviders()` in `buildMatchFoundResponse()` and `diagnoseNoMatch()` can throw `LazyInitializationException`.

Comparable read-only services (`BahmniConceptServiceImpl`, `DiseaseTemplateServiceImpl`) annotate methods with `@Transactional(readOnly = true)`. Note: if the `@Service` issue above is resolved via XML `TransactionProxyFactoryBean`, transaction management may be handled there instead.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+                                                    Date encounterDateTime, Location location,
+                                                    Set<Provider> requestedProviders) {
+        logger.debug("Querying encounters to determine no-match reason");
+
+        // Query all encounters in this visit (no encounter type filter)
+        // to determine why the matcher returned null.
+        Date startOfDay = DateUtils.truncate(encounterDateTime, Calendar.DATE);
+        logger.info("  [DIAGNOSTIC] Query params: startOfDay=" + startOfDay + ", endDate=" + encounterDateTime);
+
+        Collection<Encounter> candidates = encounterService.getEncounters(
```
Comment (sumaztwcode) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java:191]: blocking: Diagnostic query omits encounter-type filter — may produce misleading no-match reasons.

This `getEncounters()` call passes `null` for encounter types, but `EncounterSessionMatcher.findMatchingEncounter` filters by encounter type (via `EncounterTypeIdentifier`). The diagnostic may evaluate encounters of a different type than the matcher considered, returning a wrong reason (e.g., `session_expired` on a lab encounter when the matcher only looked at consultations).

Pass the same encounter type filter the matcher uses, or clearly document this as a best-effort heuristic.

File: bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java [hunk 12]
```diff
@@ -37,6 +40,7 @@ public class BahmniEncounterController extends BaseRestController {
     private EncounterTransactionMapper encounterTransactionMapper;
     private BahmniEncounterTransactionService bahmniEncounterTransactionService;
     private BahmniEncounterTransactionMapper bahmniEncounterTransactionMapper;
+    private BahmniEncounterMatchDecisionService encounterMatchDecisionService;
```
Comment (sumaztwcode) [inline bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java:44]: blocking: If OpenMRS uses the no-arg constructor, `encounterMatchDecisionService` will be null and `/match-decision` will throw NPE.

The field is only set via the multi-arg constructor. If Spring/OpenMRS uses the no-arg path (which the module loader sometimes does, then calls setters), the field stays null.

Add `@Autowired` directly on this field, or add a separate `@Autowired` setter.

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchRequest.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
+package org.openmrs.module.bahmniemrapi.encountertransaction.contract;
+
+import java.util.Date;
+
+public class EncounterMatchRequest {
+
+    private String patientUuid;
+    private String visitUuid;
+    private String providerUuid;
```
Comment (sumaztwcode) [inline bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchRequest.java:9]: suggestion: API contract deviation — `providerUuid` (singular) vs JIRA spec `providerUuids` (plural).

The JIRA tech spec defines `providerUuids` implying multi-provider support. `EncounterSessionMatcher` already accepts `Set<Provider>`. If multi-provider scenarios are needed, change to `List<String> providerUuids`. If not, document the deviation from the spec.

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchRequest.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    private String locationUuid;
+    private Date encounterDateTime;
+    private String patientProgramUuid;
```
Comment (sumaztwcode) [inline bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchRequest.java:12]: suggestion: JIRA tech spec lists `includeAll` (boolean, optional) as a request field but it is absent here.

The service hardcodes `false` for the `includeAll` parameter in the `encounterService.getEncounters()` call. Consider adding this field and threading it through, or document that voided encounters are always excluded.

File: bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java [hunk 15]
```diff
@@ -46,12 +50,37 @@ public BahmniEncounterController() {
...
+    }
+
+    @RequestMapping(method = RequestMethod.POST, value = "/match-decision")
+    @ResponseBody
+    public Map<String, Object> matchDecision(@RequestBody EncounterMatchRequest request) {
+        EncounterMatchResponse response = encounterMatchDecisionService.decideMatch(request);
+        return removeNulls(response);
+    }
+
+    private Map<String, Object> removeNulls(EncounterMatchResponse response) {
```
Comment (sumaztwcode) [inline bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java]: suggestion: Manual `removeNulls` is fragile — use `@JsonInclude(NON_NULL)` instead.

All other endpoints in this controller return typed DTOs. The project uses `@JsonInclude`/`@JsonIgnoreProperties` on DTOs like `BahmniObservation`. This approach: (a) must be updated when new fields are added to `EncounterMatchResponse`, (b) loses type safety, (c) doesn't handle nulls inside nested `Ref`/`MatchDetails`.

Annotate `EncounterMatchResponse` with `@JsonInclude(JsonInclude.Include.NON_NULL)` and return it directly.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+            } else {
+                logger.warn("  Could not resolve provider UUID: " + providerUuid);
+            }
+        } else {
+            logger.info("No provider UUID provided");
+        }
+        return null;
+    }
+
+    private Date getSearchStartDate(Date endDate) {
```
Comment (sumaztwcode) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: suggestion: `getSearchStartDate` is dead code — never called.

`diagnoseNoMatch` computes `startOfDay` via `DateUtils.truncate` directly (line 195), bypassing this method. The two approaches differ: this method uses a session-window-based start while `diagnoseNoMatch` uses a full-day window. Remove this method, or use it in `diagnoseNoMatch` for a consistent query window.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+        }
+
+        String providerUuid = request.getProviderUuid();
+        boolean providerSpecified = providerUuid != null && !providerUuid.isEmpty();
+
+        Provider provider = null;
+        if (providerSpecified) {
+            provider = resolveProvider(providerUuid);
+            if (provider == null) {
+                logger.warn("Provider not found for UUID: {}", providerUuid);
```
Comment (sumaztwcode) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java:99]: suggestion: `REASON_PROVIDER_MISMATCH` is semantically wrong here — the provider UUID doesn't exist in the system.

This conflates "provider not found" with "provider exists but differs." Compare with `INVALID_PATIENT` and `INVALID_LOCATION` which correctly return `error` status for not-found entities.

Consider `EncounterMatchResponse.error("INVALID_PROVIDER", "Provider not found.")` for consistency.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+            }
+            if (!anyMatch) {
+                logger.debug("Provider mismatch for encounter: {}", candidate.getUuid());
+                return EncounterMatchResponse.noMatch(
+                        EncounterMatchResponse.REASON_PROVIDER_MISMATCH,
+                        "Encounter exists but belongs to different provider. A new encounter will be created.");
+            }
+        }
+
+        if (location != null && candidate.getLocation() != null) {
```
Comment (sumaztwcode) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: suggestion: When `candidate.getLocation()` is null but a location was requested, the location check is silently skipped.

Execution falls to the catch-all `REASON_NO_ACTIVE_ENCOUNTER`, masking the actual issue. An encounter with no location when a specific location was requested should be treated as a location mismatch.

Add: `if (candidate.getLocation() == null) return noMatch(REASON_LOCATION_MISMATCH, ...)`

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,324 @@
...
+            if (requestedVisitLocation != null && !requestedVisitLocation.equals(encounterVisitLocation)) {
+                logger.debug("Location mismatch for encounter: {}", candidate.getUuid());
+                return EncounterMatchResponse.noMatch(
+                        EncounterMatchResponse.REASON_LOCATION_MISMATCH,
+                        "Encounter exists but is in a different location. A new encounter will be created.");
+            }
+        }
+
+        logger.debug("Encounter exists but does not match session criteria.");
+        return EncounterMatchResponse.noMatch(
```
Comment (sumaztwcode) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java:264]: suggestion: Reason `no_active_encounter` contradicts the description "Encounter exists but does not meet current session criteria."

Callers branching on `reason == "no_active_encounter"` cannot distinguish between "no encounters at all" (line 213) and this fallback where an encounter exists but didn't match any specific check. Consider a distinct reason constant like `REASON_SESSION_CRITERIA_MISMATCH`.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,324 @@
...
+import java.util.Calendar;
+import java.util.Collection;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+@Service
+public class EncounterMatchDecisionServiceImpl implements BahmniEncounterMatchDecisionService {
```
Comment (sumaztwcode) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: nit: Class should be `BahmniEncounterMatchDecisionServiceImpl` to match the interface naming convention.

All other service pairs follow: `BahmniDiagnosisService` → `BahmniDiagnosisServiceImpl`, `BahmniObsService` → `BahmniObsServiceImpl`.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,324 @@
...
+
+    private EncounterMatchResponse diagnoseNoMatch(Patient patient, Visit visit,
+                                                    Date encounterDateTime, Location location,
+                                                    Set<Provider> requestedProviders) {
+        logger.debug("Querying encounters to determine no-match reason");
+
+        // Query all encounters in this visit (no encounter type filter)
+        // to determine why the matcher returned null.
+        Date startOfDay = DateUtils.truncate(encounterDateTime, Calendar.DATE);
+        logger.info("  [DIAGNOSTIC] Query params: startOfDay=" + startOfDay + ", endDate=" + encounterDateTime);
```
Comment (sumaztwcode) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: nit: Inconsistent logging — string concatenation and `[DIAGNOSTIC]` tag.

This line and several in `resolveProvider` (lines ~296-304) use `+` concatenation instead of SLF4J `{}` parameterized format. The `[DIAGNOSTIC]` prefix at INFO level looks like debug scaffolding.

Use parameterized format and change to `logger.debug("Query params: startOfDay={}, endDate={}", startOfDay, encounterDateTime)`.

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,324 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(EncounterMatchDecisionServiceImpl.class);
+
+    private final VisitService visitService;
+    private final PatientService patientService;
+    private final LocationService locationService;
+    private final ProviderService providerService;
+    private final EncounterService encounterService;
+    private final EncounterSessionMatcher encounterSessionMatcher;
+    private final BahmniVisitLocationService bahmniVisitLocationService;
+    private final AdministrationService administrationService;
```
Comment (sumaztwcode) [inline bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: nit: `administrationService` is injected but never used in production code.

Session duration is obtained via `encounterSessionMatcher.getSessionDuration()` (line 321). This field and its constructor parameter can be removed.

**Review by sumaztwcode — CHANGES_REQUESTED** [review]
## Review for BAH-4651

Good implementation of the encounter match decision API. The core logic reuses `EncounterSessionMatcher` correctly and the diagnostic pass adds real value. Found several framework-level and logic issues to address before merge.

Found 4 blocking issues, 7 suggestions, and 4 nits.

**Acceptance Criteria**: 7/8 met (AC2 partially — uses singular `providerUuid` vs spec's plural `providerUuids`)

### Issues not in diff

| File:Line | Type | Issue |
|-----------|------|-------|
| EncounterMatchDecisionServiceImplTest.java | nit | Missing test for completely empty `EncounterMatchRequest` — `decideMatch(new EncounterMatchRequest())` should return `no_active_visit` |
| MultipleEncountersMatchException.java | question | Consider moving to `service.exception` package — currently in `matcher` package, coupling service layer to matcher internals |
| EncounterMatchResponse.java:6-16 | question | DTO contains service-layer decision vocabulary (`STATUS_*`, `REASON_*` constants) — consider moving to service interface or constants class |

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java (hunk 9 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: added in moduleApplicationContext

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java (hunk 10 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java:65]: added

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java (hunk 11 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java:191]: passing encounterType now

File: bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java (hunk 12 above)
Comment (sahupadma) [reply bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java:44]: added

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchRequest.java (hunk 13 above)
Comment (sahupadma) [reply bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchRequest.java:9]: updated the AC it should be single string only

File: bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchRequest.java (hunk 14 above)
Comment (sahupadma) [reply bahmni-emr-api/src/main/java/org/openmrs/module/bahmniemrapi/encountertransaction/contract/EncounterMatchRequest.java:12]: these are excluded

File: bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java (hunk 15 above)
Comment (sahupadma) [reply bahmnicore-omod/src/main/java/org/bahmni/module/bahmnicore/web/v1_0/controller/BahmniEncounterController.java]: removed this code

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java (hunk 16 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: removed

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java (hunk 17 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java:99]: added

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java (hunk 18 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: done

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java (hunk 19 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/BahmniEncounterMatchDecisionServiceImpl.java:264]: in match with the AC so not changing

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java (hunk 20 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: done

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java (hunk 21 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: done

File: bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java (hunk 22 above)
Comment (sahupadma) [reply bahmnicore-api/src/main/java/org/bahmni/module/bahmnicore/service/impl/EncounterMatchDecisionServiceImpl.java]: done

**Review by sumaztwcode — APPROVED** [review]
